The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11611706375/job/32333645859) [(plain)](https://github.com/rust-lang/rust/commit/5f66734bca5fd71cef5f5f01e1e6203907cc8c13/checks/32333645859/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling compiler_builtins v0.1.134
thread 'rustc' panicked at /checkout/compiler/rustc_const_eval/src/interpret/eval_context.rs:241:29:
no call frames exist
stack backtrace:
   0:     0x7f322e0cb421 - <<std[ddc22b735f97b110]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[62abb631ef17e59a]::fmt::Display>::fmt
   1:     0x7f322e127fa3 - core[62abb631ef17e59a]::fmt::write
   2:     0x7f322e0bef89 - <std[ddc22b735f97b110]::sys::pal::unix::stdio::Stderr as std[ddc22b735f97b110]::io::Write>::write_fmt
   3:     0x7f322e0cb2c2 - <std[ddc22b735f97b110]::sys::backtrace::BacktraceLock>::print
   4:     0x7f322e0cdd1b - std[ddc22b735f97b110]::panicking::default_hook::{closure#1}
   5:     0x7f322e0cda9a - std[ddc22b735f97b110]::panicking::default_hook
   6:     0x7f322980b668 - <alloc[947a67535d8e2139]::boxed::Box<rustc_driver_impl[112c401267ce0688]::install_ice_hook::{closure#0}> as core[62abb631ef17e59a]::ops::function::Fn<(&dyn for<'a, 'b> core[62abb631ef17e59a]::ops::function::Fn<(&'a std[ddc22b735f97b110]::panic::PanicHookInfo<'b>,), Output = ()> + core[62abb631ef17e59a]::marker::Sync + core[62abb631ef17e59a]::marker::Send, &std[ddc22b735f97b110]::panic::PanicHookInfo)>>::call
   7:     0x7f322e0ce578 - std[ddc22b735f97b110]::panicking::rust_panic_with_hook
   8:     0x7f322e0ce29e - std[ddc22b735f97b110]::panicking::begin_panic_handler::{closure#0}
   9:     0x7f322e0cba49 - std[ddc22b735f97b110]::sys::backtrace::__rust_end_short_backtrace::<std[ddc22b735f97b110]::panicking::begin_panic_handler::{closure#0}, !>
  11:     0x7f322976ad80 - core[62abb631ef17e59a]::panicking::panic_fmt
  12:     0x7f322976ad5b - core[62abb631ef17e59a]::option::expect_failed
  12:     0x7f322976ad5b - core[62abb631ef17e59a]::option::expect_failed
  13:     0x7f322a8c4366 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator>::eval_constant
  14:     0x7f322a8c4cb0 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_const_operand
  15:     0x7f322a8c55fc - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_assign
  16:     0x7f322a8cae37 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_statement
  17:     0x7f322a8ccf7e - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_basic_block_data
  18:     0x7f322a8c4992 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_body
  19:     0x7f322a8c14e0 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::KnownPanicsLint as rustc_mir_transform[59f325c6a6a7a434]::pass_manager::MirLint>::run_lint
  20:     0x7f322a8b8e11 - rustc_mir_transform[59f325c6a6a7a434]::pass_manager::run_passes_inner
  21:     0x7f322a8a19e1 - rustc_mir_transform[59f325c6a6a7a434]::run_analysis_to_runtime_passes
  22:     0x7f322a8a12e9 - rustc_mir_transform[59f325c6a6a7a434]::mir_drops_elaborated_and_const_checked
  23:     0x7f322bf33ae5 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>
  24:     0x7f322c0eee25 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::LocalDefId)>>::call_once
  25:     0x7f322c3c60cd - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  26:     0x7f322c142bba - rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  27:     0x7f322a87872e - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>>
  28:     0x7f322a8a05d1 - rustc_mir_transform[59f325c6a6a7a434]::mir_for_ctfe
  29:     0x7f322bf117c7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_for_ctfe::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>
  30:     0x7f322bfca759 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_for_ctfe::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::DefId)>>::call_once
  31:     0x7f322c31298f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefIdCache<rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  32:     0x7f322c1f753d - rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_for_ctfe::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7f322b4c2c89 - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::DefIdCache<rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>>
  34:     0x7f322b4642a5 - <rustc_const_eval[10a87e0838facea1]::interpret::eval_context::InterpCx<rustc_const_eval[10a87e0838facea1]::const_eval::machine::CompileTimeMachine>>::load_mir
  35:     0x7f322b4f63a4 - rustc_const_eval[10a87e0838facea1]::const_eval::eval_queries::eval_to_allocation_raw_provider
  36:     0x7f322bf267a7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  37:     0x7f322c1f15ca - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  38:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  39:     0x7f322c145e84 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7f322b4f234b - rustc_const_eval[10a87e0838facea1]::const_eval::eval_queries::eval_to_const_value_raw_provider
  41:     0x7f322bf29437 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_const_value_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  42:     0x7f322c1f235a - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_const_value_raw::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  43:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  44:     0x7f322c146014 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_const_value_raw::get_query_non_incr::__rust_end_short_backtrace
  45:     0x7f322d7468dc - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>>
  47:     0x7f322d747fb8 - <rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>::const_eval_resolve
  47:     0x7f322d747fb8 - <rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>::const_eval_resolve
  48:     0x7f322b4bc418 - <rustc_const_eval[10a87e0838facea1]::interpret::eval_context::InterpCx<rustc_const_eval[10a87e0838facea1]::const_eval::machine::CompileTimeMachine>>::push_stack_frame_raw
  49:     0x7f322b4f6e8c - rustc_const_eval[10a87e0838facea1]::const_eval::eval_queries::eval_to_allocation_raw_provider
  50:     0x7f322bf267a7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  51:     0x7f322c1f15ca - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  52:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  53:     0x7f322c145e84 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  54:     0x7f322b505cd7 - rustc_const_eval[10a87e0838facea1]::const_eval::valtrees::eval_to_valtree
  55:     0x7f3229aeaeb1 - <rustc_const_eval[10a87e0838facea1]::provide::{closure#0} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  56:     0x7f322bf16ba7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_valtree::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  57:     0x7f322c1ebcea - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_valtree::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  58:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  59:     0x7f322c1461a4 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_valtree::get_query_non_incr::__rust_end_short_backtrace
  60:     0x7f322d7468dc - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>>
  61:     0x7f322d749647 - <rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>::const_eval_global_id_for_typeck
  63:     0x7f322c9666ce - <rustc_middle[1f256af5dadb3122]::ty::consts::Const>::normalize_internal
  64:     0x7f322c9ac582 - <rustc_trait_selection[94e3061ece4172e5]::traits::normalize::AssocTypeNormalizer as rustc_type_ir[4b855d47a6f35577]::fold::TypeFolder<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>::fold_const
  65:     0x7f322cac21d2 - <rustc_middle[1f256af5dadb3122]::ty::Ty as rustc_type_ir[4b855d47a6f35577]::fold::TypeSuperFoldable<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>::try_super_fold_with::<rustc_trait_selection[94e3061ece4172e5]::traits::normalize::AssocTypeNormalizer>
  66:     0x7f322a425227 - rustc_trait_selection[94e3061ece4172e5]::traits::normalize::normalize_with_depth_to::<rustc_middle[1f256af5dadb3122]::ty::Ty>::{closure#0}
  66:     0x7f322a425227 - rustc_trait_selection[94e3061ece4172e5]::traits::normalize::normalize_with_depth_to::<rustc_middle[1f256af5dadb3122]::ty::Ty>::{closure#0}
  67:     0x7f322a41a6c1 - rustc_trait_selection[94e3061ece4172e5]::traits::normalize::normalize_with_depth::<rustc_middle[1f256af5dadb3122]::ty::Ty>
  68:     0x7f322a7a5cf8 - <rustc_infer[101194430e33c08b]::infer::at::At as rustc_trait_selection[94e3061ece4172e5]::traits::normalize::NormalizeExt>::normalize::<rustc_middle[1f256af5dadb3122]::ty::Ty>
  69:     0x7f322a79a240 - <rustc_trait_selection[94e3061ece4172e5]::traits::engine::ObligationCtxt>::normalize::<rustc_middle[1f256af5dadb3122]::ty::Ty>
  70:     0x7f322a4300bb - <rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::WfCheckingCtxt>::normalize::<rustc_middle[1f256af5dadb3122]::ty::Ty>
  71:     0x7f322a441aa9 - rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_type_defn
  72:     0x7f322a435a71 - rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_well_formed
  73:     0x7f322bf1bd95 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>
  74:     0x7f322c0eba75 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_well_formed::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::LocalDefId)>>::call_once
  75:     0x7f322c3bd484 - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  76:     0x7f322c2ac011 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  77:     0x7f322a67160b - rustc_middle[1f256af5dadb3122]::query::plumbing::query_ensure_error_guaranteed::<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, ()>
  78:     0x7f322a66049c - <rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::ParallelGuard>::run::<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir[a62d7f432d8570bc]::hir::ItemId], rustc_span[42cf5186911a921b]::ErrorGuaranteed, <rustc_middle[1f256af5dadb3122]::hir::ModuleItems>::par_items<rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}>
  79:     0x7f322a6de717 - <rustc_middle[1f256af5dadb3122]::hir::ModuleItems>::par_items::<rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_mod_type_wf::{closure#0}>
  80:     0x7f322a44819d - rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_mod_type_wf
  81:     0x7f322bf1bb95 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>
  82:     0x7f322c0eb535 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_mod_type_wf::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::LocalModDefId)>>::call_once
  83:     0x7f322c370610 - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_span[42cf5186911a921b]::def_id::LocalModDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  84:     0x7f322c180761 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  85:     0x7f322a686767 - <rustc_middle[1f256af5dadb3122]::hir::map::Map>::par_for_each_module::<rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate::{closure#0}::{closure#0}>::{closure#0}
  86:     0x7f322a660a5d - <rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::ParallelGuard>::run::<(), rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::enabled::par_for_each_in<&rustc_hir[a62d7f432d8570bc]::hir_id::OwnerId, &[rustc_hir[a62d7f432d8570bc]::hir_id::OwnerId], <rustc_middle[1f256af5dadb3122]::hir::map::Map>::par_for_each_module<rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#1}::{closure#0}>
  87:     0x7f322a79776e - <rustc_session[a89c3508fa369ab6]::session::Session>::time::<(), rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate::{closure#0}>
  88:     0x7f322a45005a - rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate
  89:     0x7f3229af7c51 - rustc_interface[34b43820e9414925]::passes::analysis
  90:     0x7f322bf36313 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>
  91:     0x7f322c0efad1 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::analysis::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, ())>>::call_once
  92:     0x7f322c31e04a - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::SingleCache<rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  93:     0x7f322c141c97 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  94:     0x7f3229789585 - <rustc_middle[1f256af5dadb3122]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}::{closure#1}::{closure#5}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
  95:     0x7f322979e98a - <rustc_interface[34b43820e9414925]::interface::Compiler>::enter::<rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}::{closure#1}, core[62abb631ef17e59a]::result::Result<core[62abb631ef17e59a]::option::Option<rustc_interface[34b43820e9414925]::queries::Linker>, rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
  96:     0x7f32298291aa - <scoped_tls[881dd76dace60bad]::ScopedKey<rustc_span[42cf5186911a921b]::SessionGlobals>>::set::<rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
  97:     0x7f32297875cc - rustc_span[42cf5186911a921b]::create_session_globals_then::<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
  98:     0x7f3229835052 - std[ddc22b735f97b110]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
  99:     0x7f322983649e - <<std[ddc22b735f97b110]::thread::Builder>::spawn_unchecked_<rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#1} as core[62abb631ef17e59a]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
 100:     0x7f322e0da0d5 - <std[ddc22b735f97b110]::sys::pal::unix::thread::Thread>::new::thread_start
 102:     0x7f3228aef850 - <unknown>
 103:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.
---
warning: the ICE couldn't be written to `/checkout/library/rustc-ice-2024-10-31T12_29_28-11374.txt`: Read-only file system (os error 30)

note: rustc 1.84.0-nightly (ce9ffc826 2024-10-31) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -Z inline-mir -Z inline-mir-preserve-debug -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C embed-bitcode=yes -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [mir_drops_elaborated_and_const_checked] elaborating drops for `num::dec2flt::decimal::<impl at core/src/num/dec2flt/decimal.rs:32:1: 32:13>::MAX_DIGITS`
#1 [mir_for_ctfe] caching mir of `num::dec2flt::decimal::<impl at core/src/num/dec2flt/decimal.rs:32:1: 32:13>::MAX_DIGITS` for CTFE
thread 'rustc' panicked at /checkout/compiler/rustc_const_eval/src/interpret/eval_context.rs:241:29:
no call frames exist
stack backtrace:
   0:     0x7f322e0cb421 - <<std[ddc22b735f97b110]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[62abb631ef17e59a]::fmt::Display>::fmt
   0:     0x7f322e0cb421 - <<std[ddc22b735f97b110]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[62abb631ef17e59a]::fmt::Display>::fmt
   1:     0x7f322e127fa3 - core[62abb631ef17e59a]::fmt::write
   2:     0x7f322e0bef89 - <std[ddc22b735f97b110]::sys::pal::unix::stdio::Stderr as std[ddc22b735f97b110]::io::Write>::write_fmt
   3:     0x7f322e0cb2c2 - <std[ddc22b735f97b110]::sys::backtrace::BacktraceLock>::print
   4:     0x7f322e0cdd1b - std[ddc22b735f97b110]::panicking::default_hook::{closure#1}
   5:     0x7f322e0cda9a - std[ddc22b735f97b110]::panicking::default_hook
   6:     0x7f322980b668 - <alloc[947a67535d8e2139]::boxed::Box<rustc_driver_impl[112c401267ce0688]::install_ice_hook::{closure#0}> as core[62abb631ef17e59a]::ops::function::Fn<(&dyn for<'a, 'b> core[62abb631ef17e59a]::ops::function::Fn<(&'a std[ddc22b735f97b110]::panic::PanicHookInfo<'b>,), Output = ()> + core[62abb631ef17e59a]::marker::Sync + core[62abb631ef17e59a]::marker::Send, &std[ddc22b735f97b110]::panic::PanicHookInfo)>>::call
   7:     0x7f322e0ce578 - std[ddc22b735f97b110]::panicking::rust_panic_with_hook
   8:     0x7f322e0ce29e - std[ddc22b735f97b110]::panicking::begin_panic_handler::{closure#0}
   9:     0x7f322e0cba49 - std[ddc22b735f97b110]::sys::backtrace::__rust_end_short_backtrace::<std[ddc22b735f97b110]::panicking::begin_panic_handler::{closure#0}, !>
  11:     0x7f322976ad80 - core[62abb631ef17e59a]::panicking::panic_fmt
  12:     0x7f322976ad5b - core[62abb631ef17e59a]::option::expect_failed
  12:     0x7f322976ad5b - core[62abb631ef17e59a]::option::expect_failed
  13:     0x7f322a8c4366 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator>::eval_constant
  14:     0x7f322a8c4cb0 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_const_operand
  15:     0x7f322a8c519b - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_assign
  16:     0x7f322a8cae37 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_statement
  17:     0x7f322a8ccf7e - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_basic_block_data
  18:     0x7f322a8c4992 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::ConstPropagator as rustc_middle[1f256af5dadb3122]::mir::visit::Visitor>::visit_body
  19:     0x7f322a8c14e0 - <rustc_mir_transform[59f325c6a6a7a434]::known_panics_lint::KnownPanicsLint as rustc_mir_transform[59f325c6a6a7a434]::pass_manager::MirLint>::run_lint
  20:     0x7f322a8b8e11 - rustc_mir_transform[59f325c6a6a7a434]::pass_manager::run_passes_inner
  21:     0x7f322a8a19e1 - rustc_mir_transform[59f325c6a6a7a434]::run_analysis_to_runtime_passes
  22:     0x7f322a8a12e9 - rustc_mir_transform[59f325c6a6a7a434]::mir_drops_elaborated_and_const_checked
  23:     0x7f322bf33ae5 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>
  24:     0x7f322c0eee25 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::LocalDefId)>>::call_once
  25:     0x7f322c3c60cd - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  26:     0x7f322c142bba - rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  27:     0x7f322a87872e - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>>
  28:     0x7f322a8a05d1 - rustc_mir_transform[59f325c6a6a7a434]::mir_for_ctfe
  29:     0x7f322bf117c7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_for_ctfe::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>
  30:     0x7f322bfca759 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_for_ctfe::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::DefId)>>::call_once
  31:     0x7f322c31298f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefIdCache<rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  32:     0x7f322c1f753d - rustc_query_impl[e77a8cc94441c9ee]::query_impl::mir_for_ctfe::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7f322b4c2c89 - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::DefIdCache<rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 8usize]>>>
  34:     0x7f322b4642a5 - <rustc_const_eval[10a87e0838facea1]::interpret::eval_context::InterpCx<rustc_const_eval[10a87e0838facea1]::const_eval::machine::CompileTimeMachine>>::load_mir
  35:     0x7f322b4f63a4 - rustc_const_eval[10a87e0838facea1]::const_eval::eval_queries::eval_to_allocation_raw_provider
  36:     0x7f322bf267a7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  37:     0x7f322c1f15ca - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  38:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  39:     0x7f322c145e84 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7f322b4f234b - rustc_const_eval[10a87e0838facea1]::const_eval::eval_queries::eval_to_const_value_raw_provider
  41:     0x7f322bf29437 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_const_value_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  42:     0x7f322c1f235a - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_const_value_raw::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  43:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  44:     0x7f322c146014 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_const_value_raw::get_query_non_incr::__rust_end_short_backtrace
  45:     0x7f322d7468dc - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>>
  47:     0x7f322d747fb8 - <rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>::const_eval_resolve
  47:     0x7f322d747fb8 - <rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>::const_eval_resolve
  48:     0x7f322b4bc418 - <rustc_const_eval[10a87e0838facea1]::interpret::eval_context::InterpCx<rustc_const_eval[10a87e0838facea1]::const_eval::machine::CompileTimeMachine>>::push_stack_frame_raw
  49:     0x7f322b4f6e8c - rustc_const_eval[10a87e0838facea1]::const_eval::eval_queries::eval_to_allocation_raw_provider
  50:     0x7f322bf267a7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  51:     0x7f322c1f15ca - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  52:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  53:     0x7f322c145e84 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  54:     0x7f322b505cd7 - rustc_const_eval[10a87e0838facea1]::const_eval::valtrees::eval_to_valtree
  55:     0x7f3229aeaeb1 - <rustc_const_eval[10a87e0838facea1]::provide::{closure#0} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  56:     0x7f322bf16ba7 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_valtree::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>
  57:     0x7f322c1ebcea - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_valtree::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>)>>::call_once
  58:     0x7f322c32e51f - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  59:     0x7f322c1461a4 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::eval_to_valtree::get_query_non_incr::__rust_end_short_backtrace
  60:     0x7f322d7468dc - rustc_middle[1f256af5dadb3122]::query::plumbing::query_get_at::<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_middle[1f256af5dadb3122]::ty::ParamEnvAnd<rustc_middle[1f256af5dadb3122]::mir::interpret::GlobalId>, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 24usize]>>>
  61:     0x7f322d749647 - <rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>::const_eval_global_id_for_typeck
  63:     0x7f322c9666ce - <rustc_middle[1f256af5dadb3122]::ty::consts::Const>::normalize_internal
  64:     0x7f322c9ac582 - <rustc_trait_selection[94e3061ece4172e5]::traits::normalize::AssocTypeNormalizer as rustc_type_ir[4b855d47a6f35577]::fold::TypeFolder<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>::fold_const
  64:     0x7f322c9ac582 - <rustc_trait_selection[94e3061ece4172e5]::traits::normalize::AssocTypeNormalizer as rustc_type_ir[4b855d47a6f35577]::fold::TypeFolder<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>::fold_const
  65:     0x7f322a421d10 - <rustc_type_ir[4b855d47a6f35577]::predicate_kind::PredicateKind<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt> as rustc_type_ir[4b855d47a6f35577]::fold::TypeFoldable<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>::try_fold_with::<rustc_trait_selection[94e3061ece4172e5]::traits::normalize::AssocTypeNormalizer>
  66:     0x7f322a41e652 - <rustc_trait_selection[94e3061ece4172e5]::traits::normalize::AssocTypeNormalizer as rustc_type_ir[4b855d47a6f35577]::fold::FallibleTypeFolder<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>::try_fold_binder::<rustc_type_ir[4b855d47a6f35577]::predicate_kind::PredicateKind<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>
  67:     0x7f322a4270f7 - <rustc_trait_selection[94e3061ece4172e5]::traits::normalize::AssocTypeNormalizer as rustc_type_ir[4b855d47a6f35577]::fold::FallibleTypeFolder<rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt>>::try_fold_predicate
  68:     0x7f322a425546 - rustc_trait_selection[94e3061ece4172e5]::traits::normalize::normalize_with_depth_to::<rustc_middle[1f256af5dadb3122]::ty::predicate::Clause>::{closure#0}
  69:     0x7f322a41b381 - rustc_trait_selection[94e3061ece4172e5]::traits::normalize::normalize_with_depth::<rustc_middle[1f256af5dadb3122]::ty::predicate::Clause>
  70:     0x7f322a79a3d1 - <rustc_trait_selection[94e3061ece4172e5]::traits::engine::ObligationCtxt<rustc_trait_selection[94e3061ece4172e5]::traits::FulfillmentError>>::normalize::<rustc_middle[1f256af5dadb3122]::ty::predicate::Clause>
  71:     0x7f322a43024b - <rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::WfCheckingCtxt>::normalize::<rustc_middle[1f256af5dadb3122]::ty::predicate::Clause>
  72:     0x7f322a4287f6 - <&mut rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses::{closure#3} as core[62abb631ef17e59a]::ops::function::FnOnce<((rustc_middle[1f256af5dadb3122]::ty::predicate::Clause, rustc_span[42cf5186911a921b]::span_encoding::Span),)>>::call_once
  73:     0x7f322a48abbb - <core[62abb631ef17e59a]::iter::adapters::chain::Chain<core[62abb631ef17e59a]::iter::adapters::flatten::FlatMap<core[62abb631ef17e59a]::iter::adapters::zip::Zip<alloc[947a67535d8e2139]::vec::into_iter::IntoIter<rustc_middle[1f256af5dadb3122]::ty::predicate::Clause>, alloc[947a67535d8e2139]::vec::into_iter::IntoIter<rustc_span[42cf5186911a921b]::span_encoding::Span>>, thin_vec[6cb0334e01547508]::ThinVec<rustc_infer[101194430e33c08b]::traits::Obligation<rustc_middle[1f256af5dadb3122]::ty::predicate::Predicate>>, rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses::{closure#4}>, core[62abb631ef17e59a]::iter::adapters::map::Map<core[62abb631ef17e59a]::iter::adapters::flatten::FlatMap<core[62abb631ef17e59a]::slice::iter::Iter<(rustc_middle[1f256af5dadb3122]::ty::predicate::Clause, rustc_span[42cf5186911a921b]::span_encoding::Span)>, core[62abb631ef17e59a]::option::Option<(rustc_middle[1f256af5dadb3122]::ty::predicate::Clause, rustc_span[42cf5186911a921b]::span_encoding::Span)>, rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses::{closure#2}>, rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses::{closure#3}>> as core[62abb631ef17e59a]::iter::traits::iterator::Iterator>::next
  74:     0x7f322a46f68b - <alloc[947a67535d8e2139]::vec::Vec<rustc_infer[101194430e33c08b]::traits::Obligation<rustc_middle[1f256af5dadb3122]::ty::predicate::Predicate>> as alloc[947a67535d8e2139]::vec::spec_from_iter::SpecFromIter<rustc_infer[101194430e33c08b]::traits::Obligation<rustc_middle[1f256af5dadb3122]::ty::predicate::Predicate>, core[62abb631ef17e59a]::iter::adapters::chain::Chain<core[62abb631ef17e59a]::iter::adapters::flatten::FlatMap<core[62abb631ef17e59a]::iter::adapters::zip::Zip<alloc[947a67535d8e2139]::vec::into_iter::IntoIter<rustc_middle[1f256af5dadb3122]::ty::predicate::Clause>, alloc[947a67535d8e2139]::vec::into_iter::IntoIter<rustc_span[42cf5186911a921b]::span_encoding::Span>>, thin_vec[6cb0334e01547508]::ThinVec<rustc_infer[101194430e33c08b]::traits::Obligation<rustc_middle[1f256af5dadb3122]::ty::predicate::Predicate>>, rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses::{closure#4}>, core[62abb631ef17e59a]::iter::adapters::map::Map<core[62abb631ef17e59a]::iter::adapters::flatten::FlatMap<core[62abb631ef17e59a]::slice::iter::Iter<(rustc_middle[1f256af5dadb3122]::ty::predicate::Clause, rustc_span[42cf5186911a921b]::span_encoding::Span)>, core[62abb631ef17e59a]::option::Option<(rustc_middle[1f256af5dadb3122]::ty::predicate::Clause, rustc_span[42cf5186911a921b]::span_encoding::Span)>, rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses::{closure#2}>, rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses::{closure#3}>>>>::from_iter
  75:     0x7f322a4514ca - rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_where_clauses
  76:     0x7f322a4363c0 - rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_well_formed
  77:     0x7f322bf1bd95 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>
  78:     0x7f322c0eba75 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_well_formed::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::LocalDefId)>>::call_once
  79:     0x7f322c3bd484 - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  80:     0x7f322c2ac011 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  81:     0x7f322a67160b - rustc_middle[1f256af5dadb3122]::query::plumbing::query_ensure_error_guaranteed::<rustc_query_system[6086fd6f53c12db4]::query::caches::VecCache<rustc_span[42cf5186911a921b]::def_id::LocalDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, ()>
  82:     0x7f322a66049c - <rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::ParallelGuard>::run::<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir[a62d7f432d8570bc]::hir::ItemId], rustc_span[42cf5186911a921b]::ErrorGuaranteed, <rustc_middle[1f256af5dadb3122]::hir::ModuleItems>::par_items<rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}>
  83:     0x7f322a6de717 - <rustc_middle[1f256af5dadb3122]::hir::ModuleItems>::par_items::<rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_mod_type_wf::{closure#0}>
  84:     0x7f322a44819d - rustc_hir_analysis[b1e85f3a3de5f89f]::check::wfcheck::check_mod_type_wf
  85:     0x7f322bf1bb95 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>
  86:     0x7f322c0eb535 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_mod_type_wf::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, rustc_span[42cf5186911a921b]::def_id::LocalModDefId)>>::call_once
  87:     0x7f322c370610 - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::DefaultCache<rustc_span[42cf5186911a921b]::def_id::LocalModDefId, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  88:     0x7f322c180761 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  89:     0x7f322a686767 - <rustc_middle[1f256af5dadb3122]::hir::map::Map>::par_for_each_module::<rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate::{closure#0}::{closure#0}>::{closure#0}
  90:     0x7f322a660a5d - <rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::ParallelGuard>::run::<(), rustc_data_structures[b82fdf3b64c6f252]::sync::parallel::enabled::par_for_each_in<&rustc_hir[a62d7f432d8570bc]::hir_id::OwnerId, &[rustc_hir[a62d7f432d8570bc]::hir_id::OwnerId], <rustc_middle[1f256af5dadb3122]::hir::map::Map>::par_for_each_module<rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#1}::{closure#0}>
  91:     0x7f322a79776e - <rustc_session[a89c3508fa369ab6]::session::Session>::time::<(), rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate::{closure#0}>
  92:     0x7f322a45005a - rustc_hir_analysis[b1e85f3a3de5f89f]::check_crate
  93:     0x7f3229af7c51 - rustc_interface[34b43820e9414925]::passes::analysis
  94:     0x7f322bf36313 - rustc_query_impl[e77a8cc94441c9ee]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e77a8cc94441c9ee]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>
  95:     0x7f322c0efad1 - <rustc_query_impl[e77a8cc94441c9ee]::query_impl::analysis::dynamic_query::{closure#2} as core[62abb631ef17e59a]::ops::function::FnOnce<(rustc_middle[1f256af5dadb3122]::ty::context::TyCtxt, ())>>::call_once
  96:     0x7f322c31e04a - rustc_query_system[6086fd6f53c12db4]::query::plumbing::try_execute_query::<rustc_query_impl[e77a8cc94441c9ee]::DynamicConfig<rustc_query_system[6086fd6f53c12db4]::query::caches::SingleCache<rustc_middle[1f256af5dadb3122]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e77a8cc94441c9ee]::plumbing::QueryCtxt, false>
  97:     0x7f322c141c97 - rustc_query_impl[e77a8cc94441c9ee]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  98:     0x7f3229789585 - <rustc_middle[1f256af5dadb3122]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}::{closure#1}::{closure#5}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
  99:     0x7f322979e98a - <rustc_interface[34b43820e9414925]::interface::Compiler>::enter::<rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}::{closure#1}, core[62abb631ef17e59a]::result::Result<core[62abb631ef17e59a]::option::Option<rustc_interface[34b43820e9414925]::queries::Linker>, rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
 100:     0x7f32298291aa - <scoped_tls[881dd76dace60bad]::ScopedKey<rustc_span[42cf5186911a921b]::SessionGlobals>>::set::<rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
 101:     0x7f32297875cc - rustc_span[42cf5186911a921b]::create_session_globals_then::<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
 102:     0x7f3229835052 - std[ddc22b735f97b110]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>
 103:     0x7f322983649e - <<std[ddc22b735f97b110]::thread::Builder>::spawn_unchecked_<rustc_interface[34b43820e9414925]::util::run_in_thread_with_globals<rustc_interface[34b43820e9414925]::util::run_in_thread_pool_with_globals<rustc_interface[34b43820e9414925]::interface::run_compiler<core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>, rustc_driver_impl[112c401267ce0688]::run_compiler::{closure#0}>::{closure#1}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[62abb631ef17e59a]::result::Result<(), rustc_span[42cf5186911a921b]::ErrorGuaranteed>>::{closure#1} as core[62abb631ef17e59a]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
 104:     0x7f322e0da0d5 - <std[ddc22b735f97b110]::sys::pal::unix::thread::Thread>::new::thread_start
 106:     0x7f3228aef850 - <unknown>
 107:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.
---
warning: the ICE couldn't be written to `/checkout/library/rustc-ice-2024-10-31T12_29_28-11374.txt`: Read-only file system (os error 30)

note: rustc 1.84.0-nightly (ce9ffc826 2024-10-31) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -Z inline-mir -Z inline-mir-preserve-debug -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C embed-bitcode=yes -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [mir_drops_elaborated_and_const_checked] elaborating drops for `mem::transmutability::<impl at core/src/mem/transmutability.rs:294:1: 294:12>::NOTHING`
#1 [mir_for_ctfe] caching mir of `mem::transmutability::<impl at core/src/mem/transmutability.rs:294:1: 294:12>::NOTHING` for CTFE
error: could not compile `core` (lib)

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc /checkout/obj/build/bootstrap/debug/rustc --crate-name core --edition=2021 core/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 --warn=unexpected_cfgs --check-cfg 'cfg(bootstrap)' --check-cfg 'cfg(no_fp_fmt_parse)' --check-cfg 'cfg(stdarch_intel_sde)' --check-cfg 'cfg(target_os, values("rtems"))' --check-cfg 'cfg(feature, values(any()))' -C debug-assertions=on --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values("debug_refcell", "optimize_for_size", "panic_immediate_abort"))' -C metadata=c9690195bf6f5d1b -C extra-filename=-c9690195bf6f5d1b --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps -Csymbol-mangling-version=v0 '--check-cfg=cfg(feature,values(any()))' -Zunstable-options '--check-cfg=cfg(bootstrap)' -Zmacro-backtrace -Csplit-debuginfo=off -Cprefer-dynamic -Zinline-mir -Zinline-mir-preserve-debug -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Cembed-bitcode=yes -Cforce-frame-pointers=yes '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' -Z binary-dep-depinfo` (exit status: 101)
  local time: Thu Oct 31 12:29:36 UTC 2024
  network time: Thu, 31 Oct 2024 12:29:36 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_const_eval/src/check_consts/check.rs
+++ b/compiler/rustc_const_eval/src/check_consts/check.rs
@@ -589,7 +589,7 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                 // Typeck only does a "non-const" check since it operates on HIR and cannot distinguish
                 // which path expressions are getting called on and which path expressions are only used
                 // as function pointers. This is required for correctness.
-                let infcx = tcx.infer_ctxt().build(body.phase.typing_mode());
+                let infcx = tcx.infer_ctxt().build(body.typing_mode(tcx));
                 let ocx = ObligationCtxt::new_with_diagnostics(&infcx);
 
                 let predicates = tcx.predicates_of(callee).instantiate(tcx, fn_args);

--- a/compiler/rustc_const_eval/src/check_consts/ops.rs
+++ b/compiler/rustc_const_eval/src/check_consts/ops.rs
@@ -116,7 +116,7 @@ impl<'tcx> NonConstOp<'tcx> for FnCallNonConst<'tcx> {
                     let obligation =
                         Obligation::new(tcx, ObligationCause::dummy(), param_env, trait_ref);
 
-                    let infcx = tcx.infer_ctxt().build(body.phase.typing_mode());
+                    let infcx = tcx.infer_ctxt().build(body.typing_mode(tcx));
                     let mut selcx = SelectionContext::new(&infcx);
                     let implsrc = selcx.select(&obligation);
 

--- a/compiler/rustc_const_eval/src/interpret/eval_context.rs
+++ b/compiler/rustc_const_eval/src/interpret/eval_context.rs
@@ -3,7 +3,7 @@ use rustc_errors::DiagCtxtHandle;
 use rustc_hir::def_id::DefId;
 use rustc_infer::infer::TyCtxtInferExt;
 use rustc_infer::infer::at::ToTrace;
-use rustc_infer::traits::ObligationCause;
+use rustc_infer::traits::{ObligationCause, Reveal};
 use rustc_middle::mir::interpret::{ErrorHandled, InvalidMetaKind, ReportedErrorInfo};
 use rustc_middle::query::TyCtxtAt;
 use rustc_middle::ty::layout::{
@@ -116,6 +116,7 @@ impl<'tcx, M: Machine<'tcx>> FnAbiOfHelpers<'tcx> for InterpCx<'tcx, M> {
 /// This test should be symmetric, as it is primarily about layout compatibility.
 pub(super) fn mir_assign_valid_types<'tcx>(
     tcx: TyCtxt<'tcx>,
+    typing_mode: TypingMode<'tcx>,
     param_env: ParamEnv<'tcx>,
     src: TyAndLayout<'tcx>,
     dest: TyAndLayout<'tcx>,
@@ -124,7 +125,7 @@ pub(super) fn mir_assign_valid_types<'tcx>(
     // all normal lifetimes are erased, higher-ranked types with their
     // late-bound lifetimes are still around and can lead to type
     // differences.
-    if util::relate_types(tcx, param_env, Variance::Covariant, src.ty, dest.ty) {
+    if util::relate_types(tcx, typing_mode, param_env, Variance::Covariant, src.ty, dest.ty) {
         // Make sure the layout is equal, too -- just to be safe. Miri really
         // needs layout equality. For performance reason we skip this check when
         // the types are equal. Equal types *can* have different layouts when
@@ -144,6 +145,7 @@ pub(super) fn mir_assign_valid_types<'tcx>(
 #[cfg_attr(not(debug_assertions), inline(always))]
 pub(super) fn from_known_layout<'tcx>(
     tcx: TyCtxtAt<'tcx>,
+    typing_mode: TypingMode<'tcx>,
     param_env: ParamEnv<'tcx>,
     known_layout: Option<TyAndLayout<'tcx>>,
     compute: impl FnOnce() -> InterpResult<'tcx, TyAndLayout<'tcx>>,
@@ -153,7 +155,13 @@ pub(super) fn from_known_layout<'tcx>(
         Some(known_layout) => {
             if cfg!(debug_assertions) {
                 let check_layout = compute()?;
-                if !mir_assign_valid_types(tcx.tcx, param_env, check_layout, known_layout) {
+                if !mir_assign_valid_types(
+                    tcx.tcx,
+                    typing_mode,
+                    param_env,
+                    check_layout,
+                    known_layout,
+                ) {
                     span_bug!(
                         tcx.span,
                         "expected type differs from actual type.\nexpected: {}\nactual: {}",
@@ -201,6 +209,11 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             memory: Memory::new(),
             recursion_limit: tcx.recursion_limit(),
         }
+    }
+
+    pub fn typing_mode(&self) -> TypingMode<'tcx> {
+        debug_assert_eq!(self.param_env.reveal(), Reveal::All);
+        TypingMode::PostAnalysis
     }
 
     /// Returns the span of the currently executed statement/terminator.
@@ -327,7 +340,7 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             return true;
         }
         // Slow path: spin up an inference context to check if these traits are sufficiently equal.
-        let infcx = self.tcx.infer_ctxt().build(TypingMode::from_param_env(self.param_env));
+        let infcx = self.tcx.infer_ctxt().build(self.typing_mode());
         let ocx = ObligationCtxt::new(&infcx);
         let cause = ObligationCause::dummy_with_span(self.cur_span());
         // equate the two trait refs after normalization

--- a/compiler/rustc_const_eval/src/interpret/operand.rs
+++ b/compiler/rustc_const_eval/src/interpret/operand.rs
@@ -773,6 +773,7 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                 )?;
             if !mir_assign_valid_types(
                 *self.tcx,
+                self.typing_mode(),
                 self.param_env,
                 self.layout_of(normalized_place_ty)?,
                 op.layout,
@@ -832,7 +833,9 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             })
         };
         let layout =
-            from_known_layout(self.tcx, self.param_env, layout, || self.layout_of(ty).into())?;
+            from_known_layout(self.tcx, self.typing_mode(), self.param_env, layout, || {
+                self.layout_of(ty).into()
+            })?;
         let imm = match val_val {
             mir::ConstValue::Indirect { alloc_id, offset } => {
                 // This is const data, no mutation allowed.

--- a/compiler/rustc_const_eval/src/interpret/place.rs
+++ b/compiler/rustc_const_eval/src/interpret/place.rs
@@ -540,6 +540,7 @@ where
                 )?;
             if !mir_assign_valid_types(
                 *self.tcx,
+                self.typing_mode(),
                 self.param_env,
                 self.layout_of(normalized_place_ty)?,
                 place.layout,
@@ -870,8 +871,13 @@ where
     ) -> InterpResult<'tcx> {
         // We do NOT compare the types for equality, because well-typed code can
         // actually "transmute" `&mut T` to `&T` in an assignment without a cast.
-        let layout_compat =
-            mir_assign_valid_types(*self.tcx, self.param_env, src.layout(), dest.layout());
+        let layout_compat = mir_assign_valid_types(
+            *self.tcx,
+            self.typing_mode(),
+            self.param_env,
+            src.layout(),
+            dest.layout(),
+        );
         if !allow_transmute && !layout_compat {
             span_bug!(
                 self.cur_span(),

--- a/compiler/rustc_const_eval/src/interpret/stack.rs
+++ b/compiler/rustc_const_eval/src/interpret/stack.rs
@@ -596,12 +596,13 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             return interp_ok(layout);
         }
 
-        let layout = from_known_layout(self.tcx, self.param_env, layout, || {
-            let local_ty = frame.body.local_decls[local].ty;
-            let local_ty =
-                self.instantiate_from_frame_and_normalize_erasing_regions(frame, local_ty)?;
-            self.layout_of(local_ty).into()
-        })?;
+        let layout =
+            from_known_layout(self.tcx, self.typing_mode(), self.param_env, layout, || {
+                let local_ty = frame.body.local_decls[local].ty;
+                let local_ty =
+                    self.instantiate_from_frame_and_normalize_erasing_regions(frame, local_ty)?;
+                self.layout_of(local_ty).into()
+            })?;
 
         // Layouts of locals are requested a lot, so we cache them.
         state.layout.set(Some(layout));

--- a/compiler/rustc_const_eval/src/util/compare_types.rs
+++ b/compiler/rustc_const_eval/src/util/compare_types.rs
@@ -8,24 +8,15 @@ use rustc_middle::traits::ObligationCause;
 use rustc_middle::ty::{ParamEnv, Ty, TyCtxt, TypingMode, Variance};
 use rustc_trait_selection::traits::ObligationCtxt;
 
-/// Returns whether the two types are equal up to subtyping.
-///
-/// This is used in case we don't know the expected subtyping direction
-/// and still want to check whether anything is broken.
-pub fn is_equal_up_to_subtyping<'tcx>(
+/// Returns whether `src` is a subtype of `dest`, i.e. `src <: dest`.
+pub fn sub_types<'tcx>(
     tcx: TyCtxt<'tcx>,
+    typing_mode: TypingMode<'tcx>,
     param_env: ParamEnv<'tcx>,
     src: Ty<'tcx>,
     dest: Ty<'tcx>,
 ) -> bool {
-    // Fast path.
-    if src == dest {
-        return true;
-    }
-
-    // Check for subtyping in either direction.
-    relate_types(tcx, param_env, Variance::Covariant, src, dest)
-        || relate_types(tcx, param_env, Variance::Covariant, dest, src)
+    relate_types(tcx, typing_mode, param_env, Variance::Covariant, src, dest)
 }
 
 /// Returns whether `src` is a subtype of `dest`, i.e. `src <: dest`.
@@ -35,6 +26,7 @@ pub fn is_equal_up_to_subtyping<'tcx>(
 /// because we want to check for type equality.
 pub fn relate_types<'tcx>(
     tcx: TyCtxt<'tcx>,
+    typing_mode: TypingMode<'tcx>,
     param_env: ParamEnv<'tcx>,
     variance: Variance,
     src: Ty<'tcx>,
@@ -45,8 +37,7 @@ pub fn relate_types<'tcx>(
     }
 
     let mut builder = tcx.infer_ctxt().ignoring_regions();
-    // FIXME(#132279): This should eventually use the already defined hidden types.
-    let infcx = builder.build(TypingMode::from_param_env(param_env));
+    let infcx = builder.build(typing_mode);
     let ocx = ObligationCtxt::new(&infcx);
     let cause = ObligationCause::dummy();
     let src = ocx.normalize(&cause, param_env, src);

--- a/compiler/rustc_const_eval/src/util/mod.rs
+++ b/compiler/rustc_const_eval/src/util/mod.rs
@@ -8,7 +8,7 @@ mod type_name;
 
 pub use self::alignment::{is_disaligned, is_within_packed};
 pub use self::check_validity_requirement::check_validity_requirement;
-pub use self::compare_types::{is_equal_up_to_subtyping, relate_types};
+pub use self::compare_types::{relate_types, sub_types};
 pub use self::type_name::type_name;
 
 /// Classify whether an operator is "left-homogeneous", i.e., the LHS has the

--- a/compiler/rustc_middle/src/mir/mod.rs
+++ b/compiler/rustc_middle/src/mir/mod.rs
@@ -39,7 +39,7 @@ use crate::ty::fold::{FallibleTypeFolder, TypeFoldable};
 use crate::ty::print::{FmtPrinter, Printer, pretty_print_const, with_no_trimmed_paths};
 use crate::ty::visit::TypeVisitableExt;
 use crate::ty::{
-    self, AdtDef, GenericArg, GenericArgsRef, Instance, InstanceKind, List, Ty, TyCtxt,
+    self, AdtDef, GenericArg, GenericArgsRef, Instance, InstanceKind, List, Ty, TyCtxt, TypingMode,
     UserTypeAnnotationIndex,
 };
 
@@ -450,6 +450,15 @@ impl<'tcx> Body<'tcx> {
     #[inline]
     pub fn basic_blocks_mut(&mut self) -> &mut IndexVec<BasicBlock, BasicBlockData<'tcx>> {
         self.basic_blocks.as_mut()
+    }
+
+    pub fn typing_mode(&self, _tcx: TyCtxt<'tcx>) -> TypingMode<'tcx> {
+        match self.phase {
+            // FIXME(#132279): the MIR is quite clearly inside of a body, so we
+            // should instead reveal opaques defined by that body here.
+            MirPhase::Built | MirPhase::Analysis(_) => TypingMode::non_body_analysis(),
+            MirPhase::Runtime(_) => TypingMode::PostAnalysis,
+        }
     }
 
     #[inline]

--- a/compiler/rustc_middle/src/mir/syntax.rs
+++ b/compiler/rustc_middle/src/mir/syntax.rs
@@ -20,9 +20,7 @@ use smallvec::SmallVec;
 use super::{BasicBlock, Const, Local, UserTypeProjection};
 use crate::mir::coverage::CoverageKind;
 use crate::ty::adjustment::PointerCoercion;
-use crate::ty::{
-    self, GenericArgsRef, List, Region, Ty, TyCtxt, TypingMode, UserTypeAnnotationIndex,
-};
+use crate::ty::{self, GenericArgsRef, List, Region, Ty, TyCtxt, UserTypeAnnotationIndex};
 
 /// Represents the "flavors" of MIR.
 ///
@@ -103,20 +101,10 @@ impl MirPhase {
         }
     }
 
-    pub fn typing_mode<'tcx>(&self) -> TypingMode<'tcx> {
-        match self {
-            // FIXME(#132279): the MIR is quite clearly inside of a body, so we
-            // should instead reveal opaques defined by that body here.
-            MirPhase::Built | MirPhase::Analysis(_) => TypingMode::non_body_analysis(),
-            MirPhase::Runtime(_) => TypingMode::PostAnalysis,
-        }
-    }
-
     pub fn param_env<'tcx>(&self, tcx: TyCtxt<'tcx>, body_def_id: DefId) -> ty::ParamEnv<'tcx> {
-        match self.typing_mode() {
-            TypingMode::Coherence => unreachable!(),
-            TypingMode::Analysis { defining_opaque_types: _ } => tcx.param_env(body_def_id),
-            TypingMode::PostAnalysis => tcx.param_env_reveal_all_normalized(body_def_id),
+        match self {
+            MirPhase::Built | MirPhase::Analysis(_) => tcx.param_env(body_def_id),
+            MirPhase::Runtime(_) => tcx.param_env_reveal_all_normalized(body_def_id),
         }
     }
 }

--- a/compiler/rustc_mir_transform/src/inline.rs
+++ b/compiler/rustc_mir_transform/src/inline.rs
@@ -244,8 +244,13 @@ impl<'tcx> Inliner<'tcx> {
         // Normally, this shouldn't be required, but trait normalization failure can create a
         // validation ICE.
         let output_type = callee_body.return_ty();
-        if !util::relate_types(self.tcx, self.param_env, ty::Covariant, output_type, destination_ty)
-        {
+        if !util::sub_types(
+            self.tcx,
+            caller_body.typing_mode(self.tcx),
+            self.param_env,
+            output_type,
+            destination_ty,
+        ) {
             trace!(?output_type, ?destination_ty);
             return Err("failed to normalize return type");
         }
@@ -275,8 +280,13 @@ impl<'tcx> Inliner<'tcx> {
                 self_arg_ty.into_iter().chain(arg_tuple_tys).zip(callee_body.args_iter())
             {
                 let input_type = callee_body.local_decls[input].ty;
-                if !util::relate_types(self.tcx, self.param_env, ty::Covariant, input_type, arg_ty)
-                {
+                if !util::sub_types(
+                    self.tcx,
+                    caller_body.typing_mode(self.tcx),
+                    self.param_env,
+                    input_type,
+                    arg_ty,
+                ) {
                     trace!(?arg_ty, ?input_type);
                     return Err("failed to normalize tuple argument type");
                 }
@@ -285,8 +295,13 @@ impl<'tcx> Inliner<'tcx> {
             for (arg, input) in args.iter().zip(callee_body.args_iter()) {
                 let input_type = callee_body.local_decls[input].ty;
                 let arg_ty = arg.node.ty(&caller_body.local_decls, self.tcx);
-                if !util::relate_types(self.tcx, self.param_env, ty::Covariant, input_type, arg_ty)
-                {
+                if !util::sub_types(
+                    self.tcx,
+                    caller_body.typing_mode(self.tcx),
+                    self.param_env,
+                    input_type,
+                    arg_ty,
+                ) {
                     trace!(?arg_ty, ?input_type);
                     return Err("failed to normalize argument type");
                 }

--- a/compiler/rustc_mir_transform/src/validate.rs
+++ b/compiler/rustc_mir_transform/src/validate.rs
@@ -20,7 +20,7 @@ use rustc_target::spec::abi::Abi;
 use rustc_trait_selection::traits::ObligationCtxt;
 use rustc_type_ir::Upcast;
 
-use crate::util::{is_within_packed, relate_types};
+use crate::util::{self, is_within_packed};
 
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 enum EdgeKind {
@@ -583,7 +583,14 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             Variance::Covariant
         };
 
-        crate::util::relate_types(self.tcx, self.param_env, variance, src, dest)
+        crate::util::relate_types(
+            self.tcx,
+            self.body.typing_mode(self.tcx),
+            self.param_env,
+            variance,
+            src,
+            dest,
+        )
     }
 
     /// Check that the given predicate definitely holds in the param-env of this MIR body.
@@ -602,7 +609,7 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             return true;
         }
 
-        let infcx = self.tcx.infer_ctxt().build(self.body.phase.typing_mode());
+        let infcx = self.tcx.infer_ctxt().build(self.body.typing_mode(self.tcx));
         let ocx = ObligationCtxt::new(&infcx);
         ocx.register_obligation(Obligation::new(
             self.tcx,
@@ -794,10 +801,10 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                 }
             }
             ProjectionElem::Subtype(ty) => {
-                if !relate_types(
+                if !util::sub_types(
                     self.tcx,
+                    self.body.typing_mode(self.tcx),
                     self.param_env,
-                    Variance::Covariant,
                     ty,
                     place_ref.ty(&self.body.local_decls, self.tcx).ty,
                 ) {

--- a/compiler/rustc_trait_selection/src/traits/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/mod.rs
@@ -411,7 +411,7 @@ pub fn normalize_param_env_or_error<'tcx>(
     debug!("normalize_param_env_or_error: elaborated-predicates={:?}", predicates);
 
     let elaborated_env = ty::ParamEnv::new(tcx.mk_clauses(&predicates), unnormalized_env.reveal());
-    if !normalize::needs_normalization(&elaborated_env, unnormalized_env.reveal()) {
+    if !elaborated_env.has_aliases() {
         return elaborated_env;
     }
 

--- a/compiler/rustc_trait_selection/src/traits/normalize.rs
+++ b/compiler/rustc_trait_selection/src/traits/normalize.rs
@@ -1,15 +1,16 @@
 //! Deeply normalize types using the old trait solver.
 
 use rustc_data_structures::stack::ensure_sufficient_stack;
-use rustc_infer::infer::InferOk;
 use rustc_infer::infer::at::At;
+use rustc_infer::infer::{InferCtxt, InferOk};
 use rustc_infer::traits::{
     FromSolverError, Normalized, Obligation, PredicateObligations, TraitEngine,
 };
 use rustc_macros::extension;
-use rustc_middle::traits::{ObligationCause, ObligationCauseCode, Reveal};
+use rustc_middle::traits::{ObligationCause, ObligationCauseCode};
 use rustc_middle::ty::{
     self, Ty, TyCtxt, TypeFoldable, TypeFolder, TypeSuperFoldable, TypeVisitable, TypeVisitableExt,
+    TypingMode,
 };
 use tracing::{debug, instrument};
 
@@ -109,16 +110,19 @@ where
 }
 
 pub(super) fn needs_normalization<'tcx, T: TypeVisitable<TyCtxt<'tcx>>>(
+    infcx: &InferCtxt<'tcx>,
+    param_env_for_debug_assertion: ty::ParamEnv<'tcx>,
     value: &T,
-    reveal: Reveal,
 ) -> bool {
     let mut flags = ty::TypeFlags::HAS_ALIAS;
 
     // Opaques are treated as rigid with `Reveal::UserFacing`,
     // so we can ignore those.
-    match reveal {
-        Reveal::UserFacing => flags.remove(ty::TypeFlags::HAS_TY_OPAQUE),
-        Reveal::All => {}
+    match infcx.typing_mode(param_env_for_debug_assertion) {
+        TypingMode::Coherence | TypingMode::Analysis { defining_opaque_types: _ } => {
+            flags.remove(ty::TypeFlags::HAS_TY_OPAQUE)
+        }
+        TypingMode::PostAnalysis => {}
     }
 
     value.has_type_flags(flags)
@@ -154,7 +158,7 @@ impl<'a, 'b, 'tcx> AssocTypeNormalizer<'a, 'b, 'tcx> {
             "Normalizing {value:?} without wrapping in a `Binder`"
         );
 
-        if !needs_normalization(&value, self.param_env.reveal()) {
+        if !needs_normalization(self.selcx.infcx, self.param_env, &value) {
             value
         } else {
             value.fold_with(self)
@@ -178,7 +182,7 @@ impl<'a, 'b, 'tcx> TypeFolder<TyCtxt<'tcx>> for AssocTypeNormalizer<'a, 'b, 'tcx
     }
 
     fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
-        if !needs_normalization(&ty, self.param_env.reveal()) {
+        if !needs_normalization(self.selcx.infcx, self.param_env, &ty) {
             return ty;
         }
 
@@ -213,10 +217,11 @@ impl<'a, 'b, 'tcx> TypeFolder<TyCtxt<'tcx>> for AssocTypeNormalizer<'a, 'b, 'tcx
         match kind {
             ty::Opaque => {
                 // Only normalize `impl Trait` outside of type inference, usually in codegen.
-                match self.param_env.reveal() {
-                    Reveal::UserFacing => ty.super_fold_with(self),
-
-                    Reveal::All => {
+                match self.selcx.infcx.typing_mode(self.param_env) {
+                    TypingMode::Coherence | TypingMode::Analysis { defining_opaque_types: _ } => {
+                        ty.super_fold_with(self)
+                    }
+                    TypingMode::PostAnalysis => {
                         let recursion_limit = self.cx().recursion_limit();
                         if !recursion_limit.value_within_limit(self.depth) {
                             self.selcx.infcx.err_ctxt().report_overflow_error(
@@ -403,7 +408,7 @@ impl<'a, 'b, 'tcx> TypeFolder<TyCtxt<'tcx>> for AssocTypeNormalizer<'a, 'b, 'tcx
     fn fold_const(&mut self, constant: ty::Const<'tcx>) -> ty::Const<'tcx> {
         let tcx = self.selcx.tcx();
         if tcx.features().generic_const_exprs()
-            || !needs_normalization(&constant, self.param_env.reveal())
+            || !needs_normalization(self.selcx.infcx, self.param_env, &constant)
         {
             constant
         } else {
@@ -420,7 +425,7 @@ impl<'a, 'b, 'tcx> TypeFolder<TyCtxt<'tcx>> for AssocTypeNormalizer<'a, 'b, 'tcx
 
     #[inline]
     fn fold_predicate(&mut self, p: ty::Predicate<'tcx>) -> ty::Predicate<'tcx> {
-        if p.allow_normalization() && needs_normalization(&p, self.param_env.reveal()) {
+        if p.allow_normalization() && needs_normalization(self.selcx.infcx, self.param_env, &p) {
             p.super_fold_with(self)
         } else {
             p

--- a/compiler/rustc_trait_selection/src/traits/query/normalize.rs
+++ b/compiler/rustc_trait_selection/src/traits/query/normalize.rs
@@ -9,7 +9,7 @@ use rustc_macros::extension;
 pub use rustc_middle::traits::query::NormalizationResult;
 use rustc_middle::ty::fold::{FallibleTypeFolder, TypeFoldable, TypeSuperFoldable};
 use rustc_middle::ty::visit::{TypeSuperVisitable, TypeVisitable, TypeVisitableExt};
-use rustc_middle::ty::{self, Ty, TyCtxt, TypeVisitor};
+use rustc_middle::ty::{self, Ty, TyCtxt, TypeVisitor, TypingMode};
 use rustc_span::DUMMY_SP;
 use tracing::{debug, info, instrument};
 
@@ -21,7 +21,7 @@ use crate::infer::canonical::OriginalQueryValues;
 use crate::infer::{InferCtxt, InferOk};
 use crate::traits::normalize::needs_normalization;
 use crate::traits::{
-    BoundVarReplacer, Normalized, ObligationCause, PlaceholderReplacer, Reveal, ScrubbedTraitError,
+    BoundVarReplacer, Normalized, ObligationCause, PlaceholderReplacer, ScrubbedTraitError,
 };
 
 #[extension(pub trait QueryNormalizeExt<'tcx>)]
@@ -89,7 +89,7 @@ impl<'a, 'tcx> At<'a, 'tcx> {
             }
         }
 
-        if !needs_normalization(&value, self.param_env.reveal()) {
+        if !needs_normalization(self.infcx, self.param_env, &value) {
             return Ok(Normalized { value, obligations: PredicateObligations::new() });
         }
 
@@ -191,7 +191,7 @@ impl<'a, 'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for QueryNormalizer<'a, 'tcx> {
 
     #[instrument(level = "debug", skip(self))]
     fn try_fold_ty(&mut self, ty: Ty<'tcx>) -> Result<Ty<'tcx>, Self::Error> {
-        if !needs_normalization(&ty, self.param_env.reveal()) {
+        if !needs_normalization(self.infcx, self.param_env, &ty) {
             return Ok(ty);
         }
 
@@ -215,10 +215,12 @@ impl<'a, 'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for QueryNormalizer<'a, 'tcx> {
         let res = match kind {
             ty::Opaque => {
                 // Only normalize `impl Trait` outside of type inference, usually in codegen.
-                match self.param_env.reveal() {
-                    Reveal::UserFacing => ty.try_super_fold_with(self)?,
+                match self.infcx.typing_mode(self.param_env) {
+                    TypingMode::Coherence | TypingMode::Analysis { defining_opaque_types: _ } => {
+                        ty.try_super_fold_with(self)?
+                    }
 
-                    Reveal::All => {
+                    TypingMode::PostAnalysis => {
                         let args = data.args.try_fold_with(self)?;
                         let recursion_limit = self.cx().recursion_limit();
 
@@ -332,7 +334,7 @@ impl<'a, 'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for QueryNormalizer<'a, 'tcx> {
         &mut self,
         constant: ty::Const<'tcx>,
     ) -> Result<ty::Const<'tcx>, Self::Error> {
-        if !needs_normalization(&constant, self.param_env.reveal()) {
+        if !needs_normalization(self.infcx, self.param_env, &constant) {
             return Ok(constant);
         }
 
@@ -351,7 +353,7 @@ impl<'a, 'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for QueryNormalizer<'a, 'tcx> {
         &mut self,
         p: ty::Predicate<'tcx>,
     ) -> Result<ty::Predicate<'tcx>, Self::Error> {
-        if p.allow_normalization() && needs_normalization(&p, self.param_env.reveal()) {
+        if p.allow_normalization() && needs_normalization(self.infcx, self.param_env, &p) {
             p.try_super_fold_with(self)
         } else {
             Ok(p)

--- a/compiler/rustc_type_ir/src/infer_ctxt.rs
+++ b/compiler/rustc_type_ir/src/infer_ctxt.rs
@@ -12,6 +12,13 @@ use crate::{self as ty, Interner};
 /// The current typing mode of an inference context. We unfortunately have some
 /// slightly different typing rules depending on the current context. See the
 /// doc comment for each variant for how and why they are used.
+///
+/// In most cases you can get the correct typing mode automically via:
+/// - `mir::Body::typing_mode`
+/// - `rustc_lint::LateContext::typing_mode`
+///
+/// If neither of these functions are available, feel free to reach out to
+/// t-types for help.
 #[derive_where(Clone, Copy, Hash, PartialEq, Eq, Debug; I: Interner)]
 #[cfg_attr(feature = "nightly", derive(TyEncodable, TyDecodable, HashStable_NoContext))]
 pub enum TypingMode<I: Interner> {

--- a/src/tools/clippy/clippy_utils/src/qualify_min_const_fn.rs
+++ b/src/tools/clippy/clippy_utils/src/qualify_min_const_fn.rs
@@ -420,7 +420,7 @@ fn is_ty_const_destruct<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, body: &Body<'tcx>
             TraitRef::new(tcx, tcx.require_lang_item(LangItem::Destruct, Some(body.span)), [ty]),
         );
 
-        let infcx = tcx.infer_ctxt().build(body.phase.typing_mode());
+        let infcx = tcx.infer_ctxt().build(body.typing_mode(tcx));
         let mut selcx = SelectionContext::new(&infcx);
         let Some(impl_src) = selcx.select(&obligation).ok().flatten() else {
             return false;

